include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+		ior = bsdf_data_decodeIoR(bsdf_data, op);
+		ior2 = matrixCompMult(ior, ior);
+		params = instr_getParameters(instr, bsdf_data);
 	}
-#ifdef OP_BUMPMAP
-	ts_flag = instr_getOpcode(instr)==OP_BUMPMAP ? false:ts_flag;
-#endif
-#endif	
-}
-//call before executing an instruction/evaluating bsdf
-void handleTwosided(inout bool ts_flag, in instr_t instr)
-{
-#ifndef NO_TWOSIDED
-	if (!ts_flag && instr_getTwosided(instr))
+
+	vec3 eval = vec3(0.0);
+	float pdf = 0.0;
+
+	if (is_bxdf && run)
 	{
-		ts_flag = true;
-		if (currInteraction.isotropic.NdotV<0.0)
+		//speculative execution
+		uint ndf = instr_getNDF(instr);
+		float a = params_getAlpha(params);
+		float a2 = a*a;
+#ifdef ALL_ISOTROPIC_BXDFS
+		float one_minus_a2 = 1.0 - a2;
+#else
+		float ay = params_getAlphaV(params);
+		float ay2 = ay*ay;
+#endif
+
+		const float eta = colorToScalar(ior[0]);
+		const float rcp_eta = 1.0 / eta;
+		const vec3 albedo = params_getReflectance(params);
+
+		const float NdotV = irr_glsl_conditionalAbsOrMax(is_bsdf, currInteraction.isotropic.NdotV, 0.0);
+
+		bool is_valid = (NdotV > FLT_MIN);
+		bool refraction = false;
+#ifdef OP_DIELECTRIC
+		if (op == OP_DIELECTRIC && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
 		{
-			flipCurrInteraction();
-
-			currBSDFParams.isotropic.NdotL = -currBSDFParams.isotropic.NdotL;
-			currBSDFParams.isotropic.NdotH = -currBSDFParams.isotropic.NdotH;
-			currBSDFParams.TdotL = -currBSDFParams.TdotL;
-			currBSDFParams.BdotL = -currBSDFParams.BdotL;
-			currBSDFParams.TdotH = -currBSDFParams.TdotH;
-			currBSDFParams.BdotH = -currBSDFParams.BdotH;
+			irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, eta, rcp_eta);
+			refraction = true;
 		}
-	}
-#ifdef OP_BUMPMAP
-	ts_flag = instr_getOpcode(instr)==OP_BUMPMAP ? false:ts_flag;
-#endif
+		else
 #endif
-}
+			microfacet = _microfacet;
 
-#ifdef GEN_CHOICE_STREAM
-void instr_remainder_and_pdf_execute(in instr_t instr, in irr_glsl_BSDFSample s)
-{
-	uint op = instr_getOpcode(instr);
+#if defined(OP_DIELECTRIC) || defined(OP_CONDUCTOR)
+		is_valid = irr_glsl_isValidVNDFMicrofacet(microfacet.isotropic, is_bsdf, refraction, s.VdotL, eta, rcp_eta);
+#endif
 
-	//speculative execution
-	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
-	params_t params = instr_getParameters(instr, bsdf_data);
-	uint ndf = instr_getNDF(instr);
-	float a = params_getAlpha(params);
-	float a2 = a*a;
-	float n = irr_glsl_alpha2_to_phong_exp(a2);
-#ifndef ALL_ISOTROPIC_BXDFS
-	float ay = params_getAlphaV(params);
-	float ay2 = ay*ay;
-	float ny = irr_glsl_alpha2_to_phong_exp(ay2);
-#endif
-	float bxdf_eval_scalar_part = 0.0;
-	float pdf = 0.0;
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+		if (op_isDiffuse(op))
+		{
+			eval = albedo * irr_glsl_oren_nayar_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, a2);
```

Comment:
use the `wo_clamps` variant

Reply:
tied up with this https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r528997756